### PR TITLE
Travis testing

### PR DESCRIPTION
**Pull request by TomMaullin**
This PR aims to implement Travis Testing for the SPM html viewer using Docker containers and Octave.

### CODE REVIEW

**Comment by TomMaullin** [comment]
Whilst creating this PR I have come across the following issue. The below command I believe should run the testing suite for the SPM viewer.

`sudo docker run -ti -v /home/tommaullin/Documents/Repos/nidmresults-spmhtml/test:/test cmaumet/octave-spm octave --no-window-system --eval "addpath('/test'); runTest()"`

This appears to run the code as required. However, Octave throws an error saying it doesn't have the import function (which is used to load the matlab unittesting framework library).

`"error: the import function is not yet implemented in Octave"`

I found a fix on stack overflow [here](https://stackoverflow.com/questions/17600696/import-function-in-octave?rq=1) but this has given no luck as it has just transformed the error to this:

`"warning: function /test/import.m shadows a core library function
error: invalid package name: matlab.unittest.TestSuite"`

**Comment by TomMaullin** [comment]
Ah I see - Thank you for clarifying this! Do you know of any example repositories than currently contain a MOdox-based testing suite?

**Comment by cmaumet** [comment]
Thank you @gllmflndn!

@TomMaullin: Does this help: https://github.com/MOdox/MOdox#use-with-travis-ci-and-shippable ?

**Comment by TomMaullin** [comment]
I am a little confused by this `foo.m` example in the `readme.md`. I will see if I can find further examples online and let you know how it goes!

**Comment by gllmflndn** [comment]
That's because I gave you a link to the wrong project, have a look at [MOxUnit](https://github.com/MOxUnit/MOxUnit) instead. Sorry!

If you were OK to use function-based unit tests instead of class-based ones, we could otherwise write a little wrapper for Octave.

**Comment by TomMaullin** [comment]
Ah - no worries! Yes, this is exactly what I was looking for - Thank you!

I see no reason why these tests couldn't be function based instead of class based... I believe, if I remember correctly, originally, I actually wrote these as functions first before plugging them into the matlab unit testing framework. I think this is why they barely make use of the class structure.

@cmaumet what do you think of this suggestion? I quite like the look of the MOxUnit tests and I don't believe it would be too much work to convert the current tests into this format.

**Comment by cmaumet** [comment]
@TomMaullin: I would vote for the option that is the easiest to implement so pretty much up to you :)

**Comment by TomMaullin** [comment]
Okay great! I'll give the MOxUnit tests a go and let you know how they go!

**Comment by gllmflndn** [comment]
The NIDM-Results export from SPM should work with Octave 4.4. Could you list the missing functions that you are trying to use? You mention contains (endsWith would be more adequate here but is not yet available either) but it is only available since R2016b so you would anyway be better off checking the file extension with fileparts/strncmp.

**Comment by cmaumet** [comment]
@TomMaullin: can you rebase on master in order to include the octave-compatibility updates from #38? And then we can check if the tests pass?

**Comment by TomMaullin** [comment]
Hi @cmaumet ,

I have now rebased as requested. I am yet to finish writing the tests though as I was waiting for the compatability branch to merge first. I will let you know once this PR is review ready!

**Comment by TomMaullin** [comment]
Hi @cmaumet , @gllmflndn ,

The Travis tests should now work! This PR is ready for review! For a history of the Travis test results see my version of this PR on my own repository [here](https://github.com/TomMaullin/nidmresults-spmhtml/pull/1). 

Apologies, normally I would rebase this branch onto master to make reviewing easier but I think if I do that the Travis test history on my branch might be reset and we won't be able to see the previous runs anymore.

One thing I wasn't sure on was whether to keep the original Matlab unittest framework version of the testing suite as well. This version was much easier to run locally (as it downloaded all of the data from neurovault inside the matlab test code rather than in a seperate bash script) and also much faster than the version in this PR. I have removed it currently but could add it back in easily if you agree with me that it might be worth keeping.

Please let me know if you have any feedback/suggestions!

**Comment by cmaumet** [comment]
Thanks @TomMaullin! I've just enabled Travis CI for this repository and created a dummy commit to trigger the tests.

File: test/bug_patches/chi2inv.m [hunk 1]
```diff
@@ -0,0 +1,71 @@
+## Copyright (C) 2012 Rik Wehbring
```
Comment (cmaumet) [inline test/bug_patches/chi2inv.m]: @TomMaullin: is it normal that you have included this chi2inv.m file in the version-control?

File: test/bug_patches/spm_file_template.m [hunk 2]
```diff
@@ -0,0 +1,204 @@
+classdef spm_file_template
```
Comment (cmaumet) [inline test/bug_patches/spm_file_template.m]: Could we patch on-the-fly as suggested by Guillaume in https://github.com/incf-nidash/nidmresults-spmhtml/pull/38#issuecomment-389840670?

File: test/bug_patches/chi2inv.m (hunk 1 above)
Comment (TomMaullin) [reply test/bug_patches/chi2inv.m]: Ah apologies... I thought I had removed this! I'll remove it now!

File: test/bug_patches/spm_file_template.m (hunk 2 above)
Comment (TomMaullin) [reply test/bug_patches/spm_file_template.m]: Oh! Right! I had misunderstood here, yes I will change this! I thought this was intended to go in the docker image eventually.

**Comment by TomMaullin** [comment]
Hi @cmaumet ,

My preference would be to keep both versions of the testing framework, both MOxUnit and matlab unittesting framework. The reason for this is that MOxUnit works on Travis whereas matlab unittesting framework does not, but while developing and making changes locally I find the matlab unittesting framework tests are much faster and convenient as they involve no setup or bug fixes.

File: test/bug_patches/spm_file_template.m (hunk 2 above)
Comment (gllmflndn) [reply test/bug_patches/spm_file_template.m]: Hopefully it's only temporary as SPM dev includes the workaround and there is a [patch submitted](https://savannah.gnu.org/bugs/?53924) for Octave.

**Comment by TomMaullin** [comment]
Hi @gllmflndn , Oh right! Yes, I hadn't thought of that - that would make life a lot easier! I'll give this a go and let you know how smoothly it goes on Octave!

**Comment by TomMaullin** [comment]
Quick update: The SPM functions worked and the `octave-statistics` dependency has been removed. Thank you for the suggestion @gllmflndn .

File: .gitignore [hunk 3]
```diff
@@ -11,6 +11,8 @@ Data/*.png
 
 # Test data (if missing will be downloaded from NeuroVault)
 test/data
+test/MOxUnit
+test/octave-statistics
```
Comment (cmaumet) [inline .gitignore]: Can we remove this from .gitignore, now that the code is not depending anymore on octave-statistics?

Comment (TomMaullin) [reply .gitignore]: Whoops, yes this is now done!

**Comment by cmaumet** [comment]
@TomMaullin: Great!! And let's keep both versions of the tests then! Could you document alternative ways to run the tests using: 2/ Octave and 3/ Octave+docker in the README? Also, do we need to put back `runTest.m` for testing in Matlab? I currently get the following error:

```
>> addpath(fullfile(fileparts(which('nidm_results_display')), 'test'))
>> runTest();
Undefined function or variable 'runTest'.
```

Once this is done, +1 to merge.

**Comment by TomMaullin** [comment]
@cmaumet , great! I will do this!

Yes, currently the Matlab unittests have been removed so the original way of calling the tests no longer works. But I will add them back in and document everything!

**Comment by cmaumet** [comment]
Thanks! +1 to merge.

**Pull request merged**